### PR TITLE
[Serve] Fix infinite backoff in `fulfill_pending_requests`

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

Currently, `fulfill_pending_requests` retries indefinitely when scheduling a request. This increments the `backoff_index`, which is used to calculate a backoff time. Eventually the `backoff_index` causes an `OverflowError`, which terminates the scheduling task.

This change logs warnings when `backoff_index` reaches 50. It also catches `OverflowErrors` and sets the `queue_len_response_deadline_s` to the `max_queue_len_response_deadline_s`.

For posterity, with the default queue length response deadline of 0.1, the `OverflowError` gets hit after the `backoff_index` becomes 1024:

```
>>> 0.1 * (2 ** 1023)
8.98846567431158e+306
>>> 0.1 * (2 ** 1024)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: int too large to convert to float
```

## Related issue number

Addresses #43964 (closed after cherry-pick).

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds a unit test to `test_pow_2_replica_scheduler.py`.

### CODE REVIEW

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py [hunk 1]
```diff
@@ -680,6 +688,17 @@ async def fulfill_pending_requests(self):
...
+                        if (
+                            request_metadata is not None
+                            and request_metadata.multiplexed_model_id
+                        ):
+                            scheduling_time_elapsed = time.time() - start_time
+                            logger.warning(
+                                "Failed to schedule request with metadata "
+                                f"{request_metadata} after {backoff_index} "
+                                f"attempts over {scheduling_time_elapsed}s."
+                            )
```
Comment (edoakes) [inline python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:701]: request_id is already in the metadata of the log message right?

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:701]: It's in the `request_metadata` but not in the log message. The log will look like this:

```
WARNING 2024-03-13 13:44:40,595 proxy 10.1.10.5 pow_2_scheduler.py:685 - Failed to schedule request with metadata RequestMetadata(request_id='63773837-28e5-42aa-890b-2ebb502b269c', endpoint='GetList', call_method='__call__', route='/', app_name='default', multiplexed_model_id='', is_streaming=True, _request_protocol=<RequestProtocol.HTTP: 'HTTP'>, grpc_context=None) after 0 attempts over 4.100799560546875e-05s.
```

Comment (edoakes) [reply python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:701]: hmm most of this info is useless. can we just include request_id in the message and multiplexed_model_id if it's present?

also please format the string to ~2 decimal places

Comment (shrekris-anyscale) [reply python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:701]: Makes sense, I updated the log to look like this:

```
WARNING 2024-03-13 14:36:23,610 proxy 10.1.10.5 5d18a6ef-ef37-4899-825b-a4e7282a565c / pow_2_scheduler.py:698 - Failed to schedule request after 0 attempts over 0.00s. Retrying. Request ID: 5d18a6ef-ef37-4899-825b-a4e7282a565c.
```

If there's a `muliplexed_model_id`, the log will also include that:

```
Failed to schedule request ... Request ID: ...Multiplexed model ID: ...
```

Comment (edoakes) [reply python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py]: thanks, much better!

would be more similar to other logs if it were like "Failed to schedule request 'request_id' ..." but either way is ok

**Pull request merged**